### ISSUE TITLE
map.addLayer 'before' parameter failing to render text in v0.35.0 (Safari retina only?)

### ISSUE THREAD

**Issue by ashleyclough** [mapbox/mapbox-gl-js#4561]
In updating to v0.35.0 I'm consistently getting problems with adding a new layer 'before' existing layers in a style:

map.addLayer(
	    	{
	            "id": "lsoa-area",
	            "type": "fill-extrusion",
	            "source": "fuel-poverty",
	            "source-layer": "fuel_poverty_by_lsoa",

	            "paint": {
		            "fill-extrusion-base": 0,
		            "fill-extrusion-height": {
		                "property": 'fuel_poor_pc',
					    "stops": [
					        [1, 100],
					        [52, 5200]
					    ]
		            },
	                "fill-extrusion-color": {
		                "property": 'fuel_poor_pc',
					    "stops": [
					        [1, "hsl(205,22%,72%)"],
					        [52, "hsl(357,97%,44%)"]
					    ]
		            },
	                "fill-extrusion-opacity": {
		                "base": 1,
		                "stops": [[9,0.9],[13,0.6]]
		            }
	            }
	        }, 'village'
	    );

Oddly it seems to be a Mac Safari issue, but on retina screens only: tested on three separate installations, latest version of Safari, all show the same issue - no text above the 3D layer.

![v35_safari_retina](https://cloud.githubusercontent.com/assets/912721/24853897/6ed8ea98-1dd4-11e7-8ee5-2fe1016e7de2.jpg)

### Expected Behaviour v0.34:
https://parallel.co.uk/3D/fuel-poverty/index_v34.html#11/52.4924/-1.9746/0/60

### Actual Behaviour v0.35 (Mac Safari retina only):
https://parallel.co.uk/3D/fuel-poverty/index_v35.html#11/52.4924/-1.9746/0/60

**Comment by ashleyclough** [comment]
Hmm, a day later and I still can't work out what is happening: sometimes the text layer shows but then fails when the page is refreshed - some sort of caching error?

**Comment by anandthakker** [comment]
Thanks for the report @ashleyclough .

I'm able to reproduce this in Safari using the provided link, but it's pretty inconsistent, and I _can't_ reproduce it with the dev tools panel open, which makes me think it's a timing issue of some sort.

**Comment by ashleyclough** [comment]
@anandthakker I've just noticed that when it fails it doesn't display any of the text - either above or below the 3D layer - which may suggest it's the original style (mapbox://styles/parallel/civnky721001c2kpgr7520ome) or the tileset that holds the location text (parallel.9b932acd) that could be causing a timing issue somewhere in 0.35?

Seems very inconsistent: sometimes it will appear on a refresh, other times not. I also managed to get it to NOT appear in a non-retina version of Safari.

**Comment by anandthakker** [comment]
@ashleyclough I'm having difficulty reproducing this on a smaller, self-contained example.  Could I ask you to try a couple of modifications to your page and see if they make any difference?

 - Comment out the code that adds a new source & layer, and see if the labels ever fail to show up.
 - Put that ^ code back, but comment out everything _after_ it, and again see if the labels fail to show up.

**Comment by ashleyclough** [comment]
@anandthakker  That's interesting: 
https://parallel.co.uk/3D/fuel-poverty/index_v35.html#11.08/52.4383/-1.9049/0/60 
with only just the base style doesn't show any text for me.

Tried a different style (mapbox://styles/parallel/civwiyjwf002w2klk4f11a96s) which uses the same tileset and it fails too.

**Comment by ashleyclough** [comment]
@anandthakker and then the text does appear after a few refreshes with no apparent change!

**Comment by ashleyclough** [comment]
@anandthakker I may have discovered more about the issue: I duplicated the style for another project and started to replace some of the svg icons I'd added into the style (I remember there'd been an issue about the size of entire icon set in earlier GL versions) and at some point in the editing the text labels returned. I've now removed a few redundant icons from the original style and the text now appears consistently (so far, at least). Were any changes made in v0.35.0 that are likely to have affected things in this way?
